Docs (#15)
1Hive/token-oracle: docs/techincal-guide.md

<br />

## External Contract Dependencies

The ABC app relies on the following external libraries.

### Audited External Contracts

These contracts have been audited by 3rd parties. Information on past Aragon audits can be found at the following locations:
- https://github.com/aragon/security-review/blob/master/past-reports.md
- https://wiki.aragon.org/association/security/

```
// Aragon app template
import "@aragon/os/contracts/apps/AragonApp.sol";
// Aragon oracle contract
import "@aragon/os/contracts/acl/IACLOracle.sol";
// ERC-20 token contract
import "@aragon/os/contracts/lib/token/ERC20.sol";
```

### External Contracts Developed By 1Hive

```
// non-audited external contracts should be copypastad into this section
// a code comment should be added explaining it's purpose
```

These contracts have not been audited by 3rd parties.

<br />

## Roles and Permissions

The Token Balance Oracle has the following roles:
```
// roles should be copypastad into this section
// a code comment should be added explaining it's purpose
```

These roles can be set to another Aragon app or an individual address. We recommend the following permissions:

> - `SET_TOKEN_ROLE` => TBD

Reviewer: ```suggestion
- `SET_TOKEN_ROLE` => Voting
```

> - `SET_MIN_BALANCE_ROLE` => TBD

Reviewer: ```suggestion
- `SET_MIN_BALANCE_ROLE` => Voting
```

function initialize(address _token, uint256 _minBalance) external onlyInit {
		require(isContract(_token), ERROR_TOKEN_NOT_CONTRACT);

		token = ERC20(_token);
		minBalance = _minBalance;

		initialized();
}
```

<br />

## setToken

`setToken` sets the token contract address to track. The address supplied must be an Ethereum contract. An event is emitted every time this parameter is set.
```
/**
* @notice Update token address to `_token`
* @param _token The new token address
*/
function setToken(address _token) external auth(SET_TOKEN_ROLE) {
		require(isContract(_token), ERROR_TOKEN_NOT_CONTRACT);
		token = ERC20(_token);

		emit TokenSet(_token);
}
```

<br />

## setMinBalance`

`setMinBalance` sets the minimum balance that is required for the Token Balance to return true. An event is emitted every time this parameter is set.
```
/**
* @notice Update minimum balance to `_minBalance`
* @param _minBalance The new minimum balance
*/
function setMinBalance(uint256 _minBalance) external auth(SET_MIN_BALANCE_ROLE) {
		minBalance = _minBalance;

		emit MinimumBalanceSet(_minBalance);
}
```

<br />

## canPerform

> `canPerform` returns a boolean if the address specified in the `_how` parameter has a balance greater than or equal to the `minBalance` set in the Token Balance oracle. `_how` is just an array that can contain data. It is expected that the data contained is the address to check. Checks are performed on the data in the [0] position of the array to make sure that it is a valid address, then the token balance of that address is checked, then a boolean is returned depending on if the token balance is greater than or equal to the `minBalance` parameter set in the Token Balance oracle.

Reviewer: I would change some parts to something like: 

```suggestion
`canPerform` returns true if the address specified in the `_how` parameter has a balance greater than or equal to the `minBalance` set in the Token Balance oracle. `_how` is just an array that can contain data. 
It is expected that the data contained is the address to check. 
Checks are performed on the data in the [0] position of the array to make sure that it is a valid address, then the token balance of that address is checked, then a boolean is returned stating whether the token balance is greater than or equal to the `minBalance` parameter set in the Token Balance oracle or not.
```
